[DDW-391] Receive screen with address sharing functionality

daniloprates opened the pull request:
This PR enables the sharing functionality in the Wallet Receive Screen.

## Screenshots

![share](https://user-images.githubusercontent.com/376611/99155205-2a44fe00-26b6-11eb-939b-c51df74eec14.gif)

![qr1](https://user-images.githubusercontent.com/1504716/99081627-e19c2080-25a1-11eb-8bca-9a49ca6bf0c0.png)
![qr2](https://user-images.githubusercontent.com/1504716/99081630-e2cd4d80-25a1-11eb-8878-6b6bf9531762.png)
![qr3](https://user-images.githubusercontent.com/1504716/99081632-e2cd4d80-25a1-11eb-9fcd-60891007b088.png)
![qrj1](https://user-images.githubusercontent.com/1504716/99081634-e365e400-25a1-11eb-8719-6a150a8669b8.png)
![qrj2](https://user-images.githubusercontent.com/1504716/99081637-e3fe7a80-25a1-11eb-9257-9347a902be2f.png)
![qrj3](https://user-images.githubusercontent.com/1504716/99081639-e3fe7a80-25a1-11eb-8676-3fdc96227e3e.png)

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)
- [ ] Test

## Test Summary

For build 15225

**New:**
1 - The PDF download button text is a little bit different than on the design. Should the correct text be “Save as PDF” or “Download as PDF”? See [on app](https://drive.google.com/file/d/1uaDRjiGRmy3KUIzOz6n8ShSQChy-9QjK/view?usp=sharing) vs [on zeplin](https://drive.google.com/file/d/1pA96MQ11H0sBfqXnXpJN3T8RlCM4dNAQ/view?usp=sharing) - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-729581556)

2 - Save as PDF/image dialog box is absent from Storybook, eg: [this one](https://drive.google.com/file/d/1iEzRT70-UZGFRBfhInybayH2hwEuZGCj/view?usp=sharing) - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-729581556)

3 - We should have transparency for used addresses (2nd and 3rd row in design) - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-729633038)

4 - Other addresses had transparency for "Share" word (4th row) - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-729633038)

5 - When we hover something we remove opacity (1st row) - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-729633038)

6 - Remove mainnet caption for mainnet app and keep only for other networks in red  - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-729592658)

**Pending:**
1 - Missing translations for “Save as a image” button and download confirmation

**Fixed:**
1 - When I exported into PDF, sometimes the number of pages were more than 1 page. See [here](https://drive.google.com/file/d/14BCodlIOY9zkRScL51bWDFnGPgJrJ6as/view?usp=sharing)  - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726878492)

2 - I'm not sure if the background of the input is correct, specially for the [yellow theme](https://drive.google.com/file/d/1Dw1QfMWiINBqz8nqwGy5Xxd43vZOz5SU/view?usp=sharing).  - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726941569)

3 - For the issue that @mioriohk mentioned on [this comment](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726878492) I found that this reproduces when no notes are inserted on the input.  - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726941569)

4 - There are some inconsistencies with [the original design](https://app.zeplin.io/project/57acfaa7436a750316161a8b/screen/5d1cf071f7ef4156ee48371c) that @mioriohk mentioned [on slack](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1605280483435300?thread_ts=1605277152.432900&cid=GGKFXSKC6) _(bold font at the beginning and the end of the address, save QR as an image button)_ Are those changes pending or the original design was modified?  - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726941569)

**Rejected:**
1 - The name of the network is displayed [in black for Mainnet](https://drive.google.com/file/d/1GAHybzGcewAoRfAOzLLvj2BxoV9wRfZC/view?usp=sharing) and i[n red for the rest of the networks](https://drive.google.com/file/d/1wkjGhIpZL-TT2RKrhI9m0M15nziA_dxe/view?usp=sharing). - [Reported here](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-728931503)

---

## Review Checklist

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

mioriohk commented:
Hi @daniloprates Issues that we've found so far on build 15175

**New issue**
1- When I exported into PDF, sometimes the number of pages were more than 1 page. See [here](https://drive.google.com/file/d/14BCodlIOY9zkRScL51bWDFnGPgJrJ6as/view?usp=sharing)

gnpf commented:
@daniloprates here's a summary for build 15175:

1 - I'm not sure if the background of the input is correct, specially for the [yellow theme]
(https://drive.google.com/file/d/1Dw1QfMWiINBqz8nqwGy5Xxd43vZOz5SU/view?usp=sharing). @a-rukin could you confirm this, please?

2 - For the issue that @mioriohk mentioned on [this comment](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726878492) I found that this reproduces when no notes are inserted on the input.

3- There are some inconsistencies with [the original design](https://app.zeplin.io/project/57acfaa7436a750316161a8b/screen/5d1cf071f7ef4156ee48371c) that @mioriohk mentioned [on slack](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1605280483435300?thread_ts=1605277152.432900&cid=GGKFXSKC6) _(bold font at the beginning and the end of the address, save QR as an image button)_ Are those changes pending or the original design was modified?

nikolaglumac reviewed (changes requested):
@daniloprates can you please implement the same notification we have for "Copy address" on the "Download PDF" action? 🙏

nikolaglumac commented:
@daniloprates one more thing - perhaps we should also prefix the PDF filename with the wallet name? To keep it consistent with the CSV export...

cc @darko-mijic

a-rukin commented:
**Issues:**
1. Font should be regular here, not light - [image](https://user-images.githubusercontent.com/2539345/99240054-fe885c00-280c-11eb-86f4-5c9e7942cece.png)
2. [Background for input field](https://user-images.githubusercontent.com/2539345/99240104-1233c280-280d-11eb-84eb-964e53c5bfd8.png) should match BG of dialog. I suggest just use transparent BG. Then you won't have issue when you use it on different BGs in one theme

daniloprates commented:
@mioriohk @gnpf @a-rukin @nikolaglumac I have fixed the remaining issues:

* Added a PDF download notification 
* Fixed the styling issues:
  * Textarea transparent bg
  * Textarea regular font
* Removed the PDF line break
* Added a new prefix for the PDF filename: `Address-{walletName}-{UTC Date}.pdf`
* Added a filename sanitization, so wallets with `/` and other reserved characters won't cause issues

I didn't add bold styling to the beg/end of addresses, because this needs to be done in the Wallet Receive page as well, which is complicated, due to the virtual lists. I'll discuss in the design call if we are still doing this, if so I'll create a separated card.

a-rukin commented:
@darko-mijic shouldn't we change the title from "cardano ada address" to "cardano address" here?

a-rukin reviewed (approved):
Looks good

nikolaglumac commented:
Great work @daniloprates 👍 
Please update the screenshots 🙏

nikolaglumac commented:
@darko-mijic please review this one 🙏

darko-mijic commented:
I've tested it briefly @nikolaglumac. Not enough to approve it.

gnpf commented:
@daniloprates For build 15204 all previous fixes look good. We're waiting for confirmation for # 3 of [this comment](https://github.com/input-output-hk/daedalus/pull/2245#issuecomment-726941569), but meanwhile, here's a small inconsistency I found:

1 - The name of the network is displayed [in black for Mainnet](https://drive.google.com/file/d/1GAHybzGcewAoRfAOzLLvj2BxoV9wRfZC/view?usp=sharing) and i[n red for the rest of the networks](https://drive.google.com/file/d/1wkjGhIpZL-TT2RKrhI9m0M15nziA_dxe/view?usp=sharing).

daniloprates commented:
@gnpf @mioriohk:

* Bold at beg/end of addresses will be done in a different card
* "Save QR as an image button" will be done here. WIP.
* > The name of the network is displayed in black for Mainnet and in red for the rest of the networks.

    This☝️ is correct, it should be red for other networks than mainnet.

daniloprates commented:
Hi @gnpf @mioriohk @nikolaglumac 

This is what I implemented today:

* Clickable row: the entire row is now clickable. The "share" button has transparency and, on hovering, gets opaque, with the exception of a used address, which keeps transparent.
* Save QR as image: it's now possible to save the QR code as an image. Like the PDF, the action closes the dialog and displays a notification. [Still waiting for the Japanese translation](https://input-output-rnd.slack.com/archives/CBECZ2K47/p1605644525122200?thread_ts=1605544065.107600&cid=CBECZ2K47).
* Small adjustments: the paragraph break in the receive screen description was removed and the "Show used" switch now has a capitalised "S".

a-rukin commented:
> This☝️ is correct, it should be red for other networks than mainnet.

if I'm not mistaken @darko-mijic wanted to remove mainnet caption for mainnet app and keep only other networks in red

a-rukin commented:
Design - https://zpl.io/2jMDeZm 
@daniloprates you understand transparency idea a bit [wrong](https://user-images.githubusercontent.com/2539345/99527849-822d7e80-29ae-11eb-9a97-a7bf12ccb00a.png). We should have transparency for used addresses (2nd and 3rd row in design) 
other addresses had transparency for "Share" word (4th row)
when we hover something we remove opacity (1st row)

nikolaglumac commented:
> 1 - The PDF download button text is a little bit different than on the design. Should the correct text be “Save as PDF” or “Download as PDF”? See [on app](https://drive.google.com/file/d/1uaDRjiGRmy3KUIzOz6n8ShSQChy-9QjK/view?usp=sharing) vs [on zeplin](https://drive.google.com/file/d/1pA96MQ11H0sBfqXnXpJN3T8RlCM4dNAQ/view?usp=sharing)

@daniloprates @mioriohk the text should be "Save as PDF". Daedalus is an "App" and not a web page...

nikolaglumac commented:
> > This☝️ is correct, it should be red for other networks than mainnet.
> 
> if I'm not mistaken @darko-mijic wanted to remove mainnet caption for mainnet app and keep only other networks in red



> Design - https://zpl.io/2jMDeZm
> @daniloprates you understand transparency idea a bit [wrong](https://user-images.githubusercontent.com/2539345/99527849-822d7e80-29ae-11eb-9a97-a7bf12ccb00a.png).
> 
> * We should have transparency for used addresses (2nd and 3rd row in design)
> * Other addresses had transparency for "Share" word (4th row)
> * When we hover something we remove opacity (1st row)

@a-rukin please add your issues to the PR's "Testing summary". We need you to use the same reporting system as QAs are...

![Screenshot 2020-11-19 at 09 38 03](https://user-images.githubusercontent.com/376611/99641749-fb7aaf00-2a4a-11eb-895d-ccd1e719e192.png)

nikolaglumac commented on package.json:
```diff
@@ -230,6 +230,7 @@
     "route-parser": "0.0.5",
     "rust-cardano-crypto": "0.2.0",
     "safe-buffer": "5.2.1",
+    "sanitize-filename": "^1.6.3",
```
@daniloprates please set fixed version here 🙏

nikolaglumac reviewed (changes requested):
@daniloprates please see the comment that I have left in the code and also fix flow issues 🙏

gnpf commented:
@daniloprates Testing summary was updated for build 15225 on the description of this PR

daniloprates commented:
@mioriohk @gnpf @a-rukin @nikolaglumac all done:

> 1 - The PDF download button text is a little bit different than on the design. Should the correct text be “Save as PDF” or “Download as PDF”?

Changed to “Save as PDF”

> 2 - Save as PDF/image dialog box is absent from Storybook
> 3 - We should have transparency for used addresses (2nd and 3rd row in design) - Reported here
> 4 - Other addresses had transparency for "Share" word (4th row) - Reported here
> 5 - When we hover something we remove opacity (1st row) - Reported here
> 6 - Remove mainnet caption for mainnet app and keep only for other networks in red - Reported here

Fixed

> 1 - Missing translations for “Save as a image” button and download confirmation

Done

a-rukin reviewed (approved):
Great!

nikolaglumac commented:
Amazing work @daniloprates 💯
